working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+```dart
+int global;
+
+@pragma('vm:entry-point')
+void foo() => global++;
+```
+
+`@shared` global variables allow developers to tackle this problem - but hidden dependency on global state might introduce hard to diagnose and debug bugs. 
+
+We propose to tackle this problem by introducing the concept of _shared isolate_: **code running in a _shared isolate_ can only access `@shared` state and not any of isolated state, an attempt to access isolated state results in a dynamic `IsolationError`**.
+
+``` dart
+int global = 0; 
+
+@shared
+int sharedGlobal = 0;
+
+void main() async {
+  global = 42;
+  sharedGlobal = 42;
+  await Isolate.run(() {
+    print(global);  // IsolationError: Can't access 'global' when running in shared isolate
+    global = 24;  // IsolationError: Can't access 'global' when running in shared isolate
+
+    print(sharedGlobal);  // => 42
+    sharedGlobal = 24;
+  }, shared: true);  
+  print(global);  // => 42
+  print(sharedGlobal);  // => 24
+}
+```
+
+### Why not compile time isolation?
+
+It is tempting to try introducing a compile time separation between functions which only access `@shared` state and functions which can access isolated state. 
+
+One obvious approach is to introduce a modifier (e.g. `@shared`) which can be applied to function declarations and impose a number of restrictions that `@shared` functions have to satisfy. These restrictions should guarantee that `@shared` functions can only access `@shared` state.
+
+```dart 
+@shared
+void foo() {
+  // ...
+}
+```
+
+* You can only pass subtypes of `Shareable` to `@shared` methods and you can only get a `Shareable` result back.
+
+  * Consequently an instance method can only be marked  `@shared` if it is declared in a `Shareable` class.
+  
+* You can't override non-`@shared` method with `@shared` method.
+* Within a `@shared` function
+  * If `f(...)` is a static function invocation then `f` must be a `@shared` function.
+  * If `o.m(...)` is an instance method invocation, then  `o` must be a subtype of `Shareable` and `m` must be `@shared` method. 
+  * If `C(...)` is a constructor invocation then `C` must be a sharable class.
+  * If `g` is a reference to a global variable then `g` must be `@shared`. 
+
+This approach seems promising on the surface, but we quickly hit issues:
+
+* It's unclear how to treat `Object` members like `toString`, `operator ==` and `get hashCode`. These can't be marked as `@shared` but should be accessible to both `@shared` and non-`@shared` code. 
+* It's unclear how to treat function expression invocations: 
+  * Function types don't encode necessary separation between `@shared` and non-`@shared` functions.
+  * Methods like `List<T>.forEach` pose challenge because they should be usable in both `@shared` and non-`@shared` contexts.
+
+This makes us think that language changes required to achieve sound compile time delineation between `@shared` and isolate worlds are too complicated to be worth it. 
```

Comment:
Naive question: could lints catch obvious dependencies on global state?

Reply:
Obvious situations can be caught - but I think it is situations which are not obvious (dependency on global state hidden deep inside of some library which you depend on) which are interesting.